# **ONLINE RETAIL OVERVIEW**

## 1. Project Overview
The notebook presents the data preparation of the Online Retail II dataset from the UCI Machine Learning Repository.
The dataset contains transactional data from a UK-based online retailer. The goal of this notebook is to clean and prepare the raw data for further analysis.

## 2. Import libraries

In [1]:
import pandas as pd
import numpy as np

## 3. Download dataset

In [2]:
!python ../scripts/download_data.py

ZIP file already exists. Skipping download.
Extraction complete.


## 4. Load the data
First, load the excel file and check the sheet names

In [3]:
raw_data_path = "../data/raw/online_retail_II.xlsx"
excel_file = pd.ExcelFile(raw_data_path)
excel_file.sheet_names

['Year 2009-2010', 'Year 2010-2011']

Create the corresponding dataframes

In [4]:
df_2009 = pd.read_excel(excel_file, sheet_name="Year 2009-2010")
print(df_2009.shape)
df_2010 = pd.read_excel(excel_file, sheet_name="Year 2010-2011")
print(df_2010.shape)

(525461, 8)
(541910, 8)


Check the number of rows and columns in each dataset (2009–2010 and 2010–2011). Since they share the same column structure, we can merge them into a single DataFrame for further analysis.

In [5]:
df_raw = pd.concat([df_2009,df_2010], ignore_index=True)
df_raw.shape

(1067371, 8)

After merging, all rows are combined and columns remain consistent across the dataset.

## 6. Review the dataset
Display information to understand the behavior of the dataset and find the columns with null values, duplicate values,
potential cleaning needs.

### 5.1 Check for duplicates

In [6]:
print(f"Number of duplicate rows: {df_raw.duplicated().sum()}")

Number of duplicate rows: 34335


### 5.2 Null values by column

In [7]:
print(df_raw.isnull().sum().sort_values(ascending=False))

Customer ID    243007
Description      4382
StockCode           0
Invoice             0
Quantity            0
InvoiceDate         0
Price               0
Country             0
dtype: int64


### 5.3 Checking Data Types

We inspect the data types to ensure each column is represented appropriately for analysis and transformation.

In [12]:
df_raw.dtypes

Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID            object
Country                object
dtype: object

In [9]:
df_raw.to_csv("../data/raw/online_retail_combined.csv", index=False)

### 5.4 Interpretation, early insights or data issues

The key columns for the analysis are: Invoice, Description, Quantity, InvoiceDate, Price, and Country.
These fields are crucial to evaluate productivity and revenue by country, as well as to analyze the quantity of products sold by month or year.
Some initial observations from the dataset:

* The full dataset contains 1,067,371 rows and 8 columns.
* The Customer ID column has a significant number of missing values, approximately 25% of the total.
* There are 34,335 duplicate rows in the data.
* The Invoice column contains values that start with a letter ("C") to indicate canceled transactions. This column may need to be split or filtered for better analysis.

These are key aspects to address during the data cleaning and transformation process.